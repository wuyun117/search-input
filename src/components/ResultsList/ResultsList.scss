@import "../styles/variables";

.ResultsList {
  background: white;
  border: 1px solid hsl(0, 0%, 80%);
  box-shadow: 0.5rem 0.5rem 1rem hsl(0, 0%, 90%);
  list-style: none;
  margin: 0;
  padding: 0;

  .ResultsList-item {
    line-height: 1.5rem;
    margin: 0;
    padding: 0.5rem 1rem;

    &:hover {
      color: #fff;
      background: #0058cc;
    }

    .ResultsList-text {
      display: inline-block;
      font-size: 1rem;
      background: none;
      border: none;
      width: 100%;
      text-align: left;
    }
  }

  .selected {
    color: #fff;
    background: #0058cc;
  }
}

.search__body--suggestions {
  position: absolute;
  width: 100%;
  top: $search-box-height;
  left: 0rem;
}
