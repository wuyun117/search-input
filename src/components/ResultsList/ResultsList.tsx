import React from "react";
import PropTypes from "prop-types";
import "./ResultsList.scss";

/**
 * <ResultsList
 *   items={[...]}
 *   onSelect={item => console.log(item.name)}
 *   className="MyResultsList"
 * />
 *
 * @prop {Array} items List of results of form { name: string, state: { abbreviation: string } }
 * @prop {Function} onSelect Callback to execute when item is selected, accepts object.
 * @prop {mixed} ... All other props will be forwarded to the container DOM node.
 */

export type Suggestion = {
  latitude?: number;
  locality?: string;
  longtitude?: number;
  name: string;
  postcode?: number;
  state: {
    abbreviation: string;
    name?: string;
  }
}

interface IResultsListProps {
  className?: string;
  items: Array<Suggestion>;
  selectedIndex: number;
  onSelect: (arg: Suggestion) => void;
  onHover: (arg: Suggestion) => void;
}

const ResultsList: React.FC<IResultsListProps> = (props: IResultsListProps) => {
  const { className, selectedIndex, onSelect, onHover, items, ...otherProps } = props;

  return (
    <ul className={"ResultsList " + (className || "")} {...otherProps} tabIndex={0}>
      {items.map(function (item: Suggestion, index: number) {
        return (
          <li
            key={"item-" + index}
            className={`ResultsList-item ${selectedIndex === index ? 'selected' : ''}`}
            onClick={() => onSelect && onSelect(item)}
            onMouseEnter={() => onHover && onHover(item)}
            role="option"
            aria-selected={index === 2 ? true : false}
          >
            <span className="ResultsList-text">
              {item.name}, {item.state.abbreviation}
            </span>
          </li>
        );
      })}
    </ul>
  );
}

ResultsList.propTypes = {
  className: PropTypes.string,
  items: PropTypes.array.isRequired,
  onSelect: PropTypes.func.isRequired,
  onHover: PropTypes.func.isRequired
}

export default ResultsList;
