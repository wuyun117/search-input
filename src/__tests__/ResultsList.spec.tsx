import React from "react";
import { render, act, fireEvent } from "@testing-library/react";

import ResultsList from "../components/ResultsList/ResultsList";

describe("Button component", () => {
  it("component render correctly", () => {
    const selectFn = jest.fn();
    const hoverFn = jest.fn();
    const { container } = render(<ResultsList items={items} selectedIndex={2} onSelect={selectFn} onHover={hoverFn}/>);
    expect(container).toMatchSnapshot();
  });

  it("should have 5 suggestions", () => {
    const selectFn = jest.fn();
    const hoverFn = jest.fn();
    const { getAllByRole,  getByText} = render(<ResultsList items={items} selectedIndex={2} onSelect={selectFn} onHover={hoverFn}/>);
    expect(getAllByRole('option')).toHaveLength(5);
    expect(getByText('Sydney South, NSW')).toBeInTheDocument();
    expect(getByText('Sydney, NSW')).toBeInTheDocument();
    expect(getByText('Sydney International Airport, NSW')).toBeInTheDocument();
    expect(getByText('Sydney Domestic Airport, NSW')).toBeInTheDocument();
    expect(getByText('Sydenham, VIC')).toBeInTheDocument();
  });

  it("should trigger onSelect event", async () => {
    const selectFn = jest.fn();
    const hoverFn = jest.fn();
    const item = items[0];
    const { getAllByRole} = render(<ResultsList items={items} selectedIndex={2} onSelect={selectFn} onHover={hoverFn}/>);
    await act(async () => {
      fireEvent.click(getAllByRole('option')[0], {
        item
      });
    });
    expect(selectFn).toHaveBeenCalledWith(item);
  });

  it("should trigger onMouseEnter event", async () => {
    const selectFn = jest.fn();
    const hoverFn = jest.fn();
    const item = items[0];
    const { getAllByRole} = render(<ResultsList items={items} selectedIndex={2} onSelect={selectFn} onHover={hoverFn}/>);
    await act(async () => {
      fireEvent.mouseEnter(getAllByRole('option')[0], {
        item
      });
    });
    expect(hoverFn).toHaveBeenCalledWith(item);
  });
});

const items = [
  { name: "Sydney South", state: { abbreviation: "NSW" } },
  { name: "Sydney", state: { abbreviation: "NSW" } },
  { name: "Sydney International Airport", state: { abbreviation: "NSW" } },
  { name: "Sydney Domestic Airport", state: { abbreviation: "NSW" } },
  { name: "Sydenham", state: { abbreviation: "VIC" } }
];