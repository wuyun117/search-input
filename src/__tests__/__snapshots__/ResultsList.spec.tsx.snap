// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component component render correctly 1`] = `
<div>
  <ul
    class="ResultsList "
    tabindex="0"
  >
    <li
      aria-selected="false"
      class="ResultsList-item "
      role="option"
    >
      <span
        class="ResultsList-text"
      >
        Sydney South
        , 
        NSW
      </span>
    </li>
    <li
      aria-selected="false"
      class="ResultsList-item "
      role="option"
    >
      <span
        class="ResultsList-text"
      >
        Sydney
        , 
        NSW
      </span>
    </li>
    <li
      aria-selected="true"
      class="ResultsList-item selected"
      role="option"
    >
      <span
        class="ResultsList-text"
      >
        Sydney International Airport
        , 
        NSW
      </span>
    </li>
    <li
      aria-selected="false"
      class="ResultsList-item "
      role="option"
    >
      <span
        class="ResultsList-text"
      >
        Sydney Domestic Airport
        , 
        NSW
      </span>
    </li>
    <li
      aria-selected="false"
      class="ResultsList-item "
      role="option"
    >
      <span
        class="ResultsList-text"
      >
        Sydenham
        , 
        VIC
      </span>
    </li>
  </ul>
</div>
`;
